src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -115,48 +128,73 @@ library GPv2Encoding {
+        // which is required by EIP-712 as a prefix to the order data.
+        // Furthermore structs are not packed in Solidity, and there is the
+        // order type hash prefix followed by the 9 order fields to hash for a
+        // total of `10 * sizeof(uint) = 320` bytes.
+        bytes32 orderDigest;
         // solhint-disable-next-line no-inline-assembly
         assembly {
-            // NOTE: Structs are not packed in Solidity, and there are 9 order
-            // fields to hash for a total of `9 * sizeof(uint) = 288` bytes.
-            orderDigest := keccak256(order, 288)
+            mstore(order, ORDER_TYPE_HASH)
+            orderDigest := keccak256(order, 320)
         }
 
-        // NOTE: Now compute the digest that was used for signing. This is not
-        // the same as the order digest as it includes a message signature
-        // prefix as well as the domain separator. This is done using a scratch
-        // region of Solidity memory past the last allocation (which is stored
-        // at `0x40` memory address).
+        // NOTE: Solidity allocates, but does not free, memory when calling the
+        // ABI encoding methods as well as the `ecrecover` precompile. However,
+        // we can restore the free memory pointer to before we made allocations
+        // to effectively free the memory.
+        // <https://solidity.readthedocs.io/en/v0.7.4/internals/layout_in_memory.html>
+        uint256 freeMemoryPointer;
         // solhint-disable-next-line no-inline-assembly
         assembly {
-            let scratch := mload(0x40)
-
-            // NOTE: Message prefix is right-padded with 0's so that the first
-            // byte of the message prefix is at `scratch` memory location. 4 0's
-            // are needed as the message prefix is exactly 28 bytes long.
-            // Consequently, the two hashes are stored at 28 and 60 bytes from
-            // the start of the message, and the total length is 92 bytes.
-            mstore(scratch, "\x19Ethereum Signed Message:\n64\x00\x00\x00\x00")
-            mstore(add(scratch, 28), domainSeparator)
-            mstore(add(scratch, 60), orderDigest)
-            signingDigest := keccak256(scratch, 92)
+            freeMemoryPointer := mload(0x40)
+        }
+
+        bytes32 signingDigest;
+        if (v & 0x80 == 0) {
+            // NOTE: The order is signed using the EIP-712 sheme, the signing
+            // hash is `"\x19\x01" || domainSeparator || orderDigest`.
+            signingDigest = keccak256(
+                abi.encodePacked("\x19\x01", domainSeparator, orderDigest)
+            );
+        } else {
+            // NOTE: The order is signed using generic message scheme, prefixed
+            // with `"\x19Ethereum Signed Message:\n64"` where 64 is length of
+            // the message being signed, which for Gnosis Protocol is
+            // `domainSeparator || orderDigest`.
+            signingDigest = keccak256(
```

Comment:
Are you following any standard here? just wondering because I see that this is done differently everywhere. Not sure if something like this is in the conditional tokens, but we have a different logic in the safe, 0x has something like this too and there is actually an EIP draft with another logic (which is inactive since january: https://ethereum-magicians.org/t/erc-2126-signature-type-recognition/3392/19). 

Again this is not really relevant for functionality (and I assume this solution was used to save gas, as it saves 1 time memory copy).

What feels weird is that depending on your signing your digest for a order is different ... I would not have expected this.

Reply:
> Are you following any standard here?

The standard for `eth_sign` is `"\x19Ethereum Signed Message:\n64" || data.length.toString(radix: 10) || data`. We just chose data to be `domainSeparator || orderStructHash`.

> an EIP draft with another logic

This EIP was closed without merging, so I don't think we should use it. ALso it seams related to differentiating different signing schemes. Here, we do something similar to the Safe where we encode in the signature `v` parameter which signing scheme is being used, and we support `eth_sign` and EIP-712.

> What feels weird is that depending on your signing your digest for a order is different

 The order digest is the same for all signature schemes, it is the EIP-712 `Order` struct hash. The signature hash, however, is different per scheme, but this is expected as different signing schemes use different signing hashes. For EIP-712 its `"\x19\x01" || domainSeparator || structHash`, while for `eth_sign` the signature hash needs to be prefixed by "Ethereum Signed Message" (see :point_up:).

The reason we want to support more than one signature type is:
- `eth_sign` is required as not all wallets support EIP-712, so we want to make sure we can recover signing addresses for those wallets.
- EIP-712 can provide really nice UX and display all order parameters that are being signed (instead of just random 64 bytes).

Reply:
> Are you following any standard here?

Ok maybe I should have phrased it differently :P 

Why did you choose to use as the `data` the concat of `domainSeperator` and `orderStructHash`? And should we aim to use the same logic in Gnosis products, e.g. same signature format in Safe and GP.

> What feels weird is that depending on your signing your digest for a order is different

Still an order is represented by different digests in your code ... but I think that is where my thinking here is "wrong" ... in the Safe the `digest` is the Safe transaction hash and uniquely represents a transaction. The `digest` here is just a variable saying what was signed. So in the Safe we generate a hash and sign it, either raw or with eth_sign. While you use different data depending if you use raw signing (that would be eip721) vs when you use eth_sign.

As a unrelated side note to the signing, some wallets present the user what they sign for eth_sign, but that never includes the message prefix. Ledger also hashes that value again with sha256 (whyever). Might be interesting for validation purposes in the interface in the future.

Reply:
> Why did you choose to use as the data the concat of domainSeperator and orderStructHash?

I see. So, its for a few reasons:
- We needed to make mix in domain replay protection, so we re-use the EIP-712 domain separator for that (it includes chain ID and contract address in the `GPv2Settlement` contract).
- We need to compute an order hash to check the order wasn't used in a previous batch, so we need to compute `orderDigest` (AKA the order struct hash) already anyway.
- I thought the symmetry in both signing schemes was nice, where only the prefix was different between the two.

> The `digest` here is just a variable saying what was signed.

Correct, the `signingDigest` is precisely that. Note that `order.digest` is unique per order data and different than the `signingDigest`.

> some wallets present the user what they sign for eth_sign, but that never includes the message prefix

The main issue is that you will just see "you are signing these bytes" but without providing any context as to where they came from, which is why EIP-712 is nice.

> Ledger also hashes that value again with sha256 (whyever). Might be interesting for validation purposes in the interface in the future.

Hmm, that might be a problem then...

Reply:
Created https://github.com/gnosis/oba-contracts/issues/155 to explore @rmeissner's idea. I'd like to keep it separate from this PR though for now though.